[Webhook] Added component documentation for use in combination with Mailer (#19249)
symfony/symfony-docs: webhook.rst

Webhook
=======

.. versionadded:: 6.3

    The Webhook component was introduced in Symfony 6.3

The Webhook component is used to respond to remote webhooks to trigger actions
in your application. This document focuses on using webhooks to listen to remote
events in other Symfony components.

Installation
------------

.. code-block:: terminal

    $ composer require symfony/webhook

Usage in combination with the Mailer component
----------------------------------------------

When using a third-party mailer, you can use the Webhook component to receive
webhook calls from the third-party mailer.

In this example Mailgun is used with ``'mailer_mailgun'`` as webhook type.
Any type name can be used as long as it's unique. Make sure to use it in the

> routing configuration, the webhook URL and the Remote Event Consumer.

Reviewer: ```suggestion
routing configuration, the webhook URL and the RemoteEvent consumer.
```

> Install the third party mailer as described in the documentation of the mailer

Reviewer: We may can add a link to Mailer documentation
Edit : And in Mailer documentation, maybe mention this new page too :)

Author: Indeed. I also want to link from the Mailer documentation to the webhook as well.

Author: Can I link to the third party bit directly? If so, how?

Reviewer 2: As we don't have much info for a third party bridge in the docs we could link to the overview or link to the README in the code repo.

What would you prefer? 

Btw, WDYT about this PR? https://github.com/symfony/symfony-docs/pull/19263

Author: That PR would be a good addition, @OskarStark. Nice.

As for the link, see my comment below. I'm not sure if that works as I hope it works. 😅

> :doc:`Mailer component </mailer>`.

Author: ```suggestion
:doc:`Mailer component </mailer_using-a-3rd-party-transport>`.
```

@OskarStark would this notation work to link directly to that part of the documentation?

Reviewer: You should add a target on Mailer documentation page right before the title of the section like `.. _mailer-using-3rd-party-transport:` as it's done for example with Transport Setup on top of the page
Then you can use this in a ref directive

See [documentation format](https://symfony.com/doc/current/contributing/documentation/format.html#adding-links) for more infos

Author: Thanks! Updated accordingly.

Webhook
=======

.. versionadded:: 6.3

    The Webhook component was introduced in Symfony 6.3

> The webhook component is used to respond to remote webhooks to trigger actions

Reviewer: ```suggestion
The Webhook component is used to respond to remote webhooks to trigger actions
```

> in your application. This document focusses on using webhooks to listen to remote

Reviewer: ```suggestion
in your application. This document focuses on using webhooks to listen to remote
```

events in other Symfony components.

Installation
------------

.. code-block:: terminal

    $ composer require symfony/webhook

Usage in combination with the Mailer component
----------------------------------------------

> When using a third-party mailer, you can use the webhook component to receive

Reviewer: ```suggestion
When using a third-party mailer, you can use the Webhook component to receive
```

webhook calls from the third-party mailer.

In this example Mailgun is used with ``'mailer_mailgun'`` as webhook type.
Any type name can be used as long as it's unique. Make sure to use it in the
routing configuration, the webhook URL and the Remote Event Consumer.

Install the third party mailer as described in the documentation of the mailer
component.

> The webhook component routing needs to be defined:

Reviewer: ```suggestion
The Webhook component routing needs to be defined:
```

                        <framework:service>mailer.webhook.request_parser.mailgun</framework:service>
                        <framework:secret>%env(MAILER_MAILGUN_SECRET)%</framework:secret>
                    </framework:routing>
                </framework:webhook>
            </framework:config>
        </container>

    .. code-block:: php

        // config/packages/framework.php
        use App\Webhook\MailerWebhookParser;
        use Symfony\Config\FrameworkConfig;
        return static function (FrameworkConfig $frameworkConfig): void {
            $webhookConfig = $frameworkConfig->webhook();
            $webhookConfig
                ->routing('mailer_mailgun')
                ->service('mailer.webhook.request_parser.mailgun')
                ->secret('%env(MAILER_MAILGUN_SECRET)%')
            ;
        };

Currently, the following third-party mailer services support webhooks:

=============== ==========================================
Mailer service  Parser service name
=============== ==========================================
Mailgun         ``mailer.webhook.request_parser.mailgun``
Postmark        ``mailer.webhook.request_parser.postmark``

> Sendgrid        ``mailer.webhook.request_parser.sendgrid``

Reviewer: IIIC Sendgrid was added in 6.4

Reviewer 2: If this is true we should only list 6.3 ones. And add Sendgrid in 6.4 after merge with a versionadded directive

Author: You're right. 6.3 only has Mailgun and Postmark.
6.4 added Sendgrid.

I removed Sendgrid for now. I can add it in a separate PR for 6.4.

=============== ==========================================

Set up the webhook in the third-party mailer. For Mailgun, you can do this
in the control panel. As URL, make sure to use the ``/webhook/mailer_mailgun``
path behind the domain you're using.

Mailgun will provide a secret for the webhook. Add this secret to your ``.env``
file:

.. code-block:: env

    MAILER_MAILGUN_SECRET=your_secret

> With this done, you can now add Remote Event Consumers to react to the webhooks::

Reviewer: ```suggestion
With this done, you can now add RemoteEvent Consumer to react to the webhooks::
```
Not sure about space and/or capitalize for consumer

Reviewer 2: a RemoteEvent consumer
